src/s3-backend.js

```diff
@@ -0,0 +1,310 @@
+let storageBackend = require('./storage-backend.js');
+let encodeURL = storageBackend.encodeURL;
+let decodeURL = storageBackend.decodeURL;
+let StorageBackend = storageBackend.StorageBackend;
+let contentDisposition = require('content-disposition');
+let url = require('url');
+let assert = require('assert');
+let debug = require('debug')('cloud-mirror:s3backend');
+let stream = require('stream');
+let cookie = require('cookie');
+
+/**
+ * We use this to wrap the upload object returned by s3.upload so that we get a
+ * promise interface.  Since this api method does not return the standard
+ * AWS.Result class, it's not wrapped in Promises by the aws-sdk-promise
+ * wrapper.  Maybe we should consider adding this wrapper to the
+ * aws-sdk-promise class...
+ */
+let wrapSend = (upload) => {
+  return new Promise((res, rej) => {
+    // TODO: Make this configurable?
+    let abortTimer = setTimeout(upload.abort.bind(upload), 1000 * 60 * 60);
+    debug('initiating upload');
+    upload.send((err, data) => {
+      clearTimeout(abortTimer);
+      if (err) {
+        debug('upload error');
+        debug(err.stack || err);
+        rej(err);
+      } else {
+        debug('upload completed');
+        res(data);
+      }
+    });
+  });
+};
+
+/**
+ * Implementation of the StorageBackend class for S3.  This class will put all
+ * cached files into the same bucket.
+ */
+class S3Backend extends StorageBackend {
+
+  constructor(config) {
+    assert(config.region, 'must specify aws s3 region');
+    let superConfig = {
+      memcachedTTL: config.memcachedTTL,
+      sqs: config.sqs,
+      memcached: config.memcached,
+      allowedPatterns: config.allowedPatterns,
+      id: 'S3_' + config.region,
+    };
+    super(superConfig);
```

Comment:
nit but can we just do: `super({....})` and avoid the variable...

Also shouldn't `memcachedTTL` be `Math.max(memcachedTTL, 24 * 60 * 60 * 1000 * lifeCycleDays)` or something like that... as the S3 backend is configured with some amount if lifeCycle days right?